ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fullsync/BetterSyncTargetEvaluator.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.fullsync;
+
+import tech.pegasys.pantheon.ethereum.eth.manager.ChainState;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeer;
+import tech.pegasys.pantheon.ethereum.eth.manager.EthPeers;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.util.uint.UInt256;
+
+import java.util.Optional;
+
+public class BetterSyncTargetEvaluator {
+
+  private final SynchronizerConfiguration config;
+  private final EthPeers ethPeers;
+
+  public BetterSyncTargetEvaluator(
+      final SynchronizerConfiguration config, final EthPeers ethPeers) {
+    this.config = config;
+    this.ethPeers = ethPeers;
+  }
+
+  public boolean shouldSwitchSyncTarget(final EthPeer currentPeer) {
+    final ChainState currentPeerChainState = currentPeer.chainState();
+    final Optional<EthPeer> maybeBestPeer = ethPeers.bestPeer();
+
+    return maybeBestPeer
+        .map(
+            bestPeer -> {
+              if (EthPeers.BEST_CHAIN.compare(bestPeer, currentPeer) <= 0) {
+                // Our current target is better or equal to the best peer
+                return false;
+              }
+              // Require some threshold to be exceeded before switching targets to keep some
+              // stability when multiple peers are in range of each other
+              final ChainState bestPeerChainState = bestPeer.chainState();
+              final UInt256 tdDifference =
+                  bestPeerChainState
+                      .getBestBlock()
+                      .getTotalDifficulty()
+                      .minus(currentPeerChainState.getBestBlock().getTotalDifficulty());
+              if (tdDifference.compareTo(config.downloaderChangeTargetThresholdByTd()) > 0) {
```

Comment:
The defaults we have for these values don't seem to make much sense.  Looks like mainnet is changing by ~10^15 per block but the default threshold is 10^9.  And td is tricky because the rate of change should depend on the particular chain.  In any case, we should probably try to pick better defaults as part of this PR while we're here.

Reply:
That's a particularly astute yet inconvenient observation... 

The situation we want to avoid is where our first peer has a very long but low difficulty chain, trying to draw us into a tar pit and spending a long time sync'ing their chain. Currently if their chain height was higher we'd continue syncing to them until we reached their head and only then switch to the actual best peer and begin syncing the right chain.

That said, it takes quite a lot of effort to generate a really long PoW chain because the difficulty will always keep blocks around 15s apart. 

I'm inclined to set it to a high value like 10^18 which means the default is almost always switch on chain height before the TD threshold was reached. It would still differentiate between ETH and ETC though we have an explicit Dao block check for that as well now. Having the setting around is potentially quite useful for cases like when the Ropsten fork went wrong - you'd want to set it to something much smaller to ensure you switched to the best chain faster.

Reply:
Sounds reasonable.  The height threshold is pretty low as well - should we bump that up?